overlay_new.py

```diff
@@ -0,0 +1,197 @@
+        visible. """
+
+    text_size = 1
+    spacing = 10
+    _, text_height = Canvas.get_text_dimensions("blah", text_size)
+    chars_per_line = 75
+
+    def draw_base(self, canvas: Canvas):
+        pass
+
+    def draw_data(self, canvas: Canvas, data: Data):
+        if not data.has_message():
+            return
+        message = data.get_message()
+        display_str = f"Message: {message}"
+
+        # Display each line of the message
+        line_y_coord = Message.spacing + Message.text_height
+        for line in wrap(display_str, Message.chars_per_line):
+            coord = (Message.spacing, line_y_coord)
+            canvas.draw_text(line, coord, Message.text_size)
+            line_y_coord += Message.text_height + Message.spacing
+
+class OverlayNew(Overlay):
+
+    def __init__(self, bike=None):
+        super(OverlayNew, self).__init__(bike)
+
+        self.start_time = time()
+
+        # Generate coordinates for each of the data fields in the bottom corners.
+        # Note that these coordinates are for the bottom left corner of each field.
+        spacing = 20
+        row_coords = [
+            self.height - (2 * spacing + DataField.height),
+            self.height - spacing,
+        ]
+        col_coords = [
+            spacing,
+            2 * spacing + DataField.width,
+            self.width - 2 * (spacing + DataField.width),
+            self.width - (spacing + DataField.width),
+        ]
+        # This lambda returns the coordinates of the data field in column x, row y
+        data_field_coord = lambda x, y: (col_coords[x], row_coords[y])
+
+        # Create all drawable overlay objects
+        self.drawables = [
+            SpeedField(data_field_coord(0, 0)),
+            DataField("TIME", self.time_func, data_field_coord(0, 1)),
+            DataField("RPM", self.get_data_func("cadence"), data_field_coord(1, 0)),
+            DataField("BPM", self.get_data_func("heartRate"), data_field_coord(1, 1)),
+
+            DataField("REC KPH", self.get_data_func("rec_speed", 1), data_field_coord(2, 0)),
+            DataField("ZONE KM", self.get_data_func("zdist", 2, 0.001), data_field_coord(2, 1)),
+            DataField("MAX KPH", self.get_data_func("predicted_max_speed", 1), data_field_coord(3, 0)),
+            DataField("DIST KM", self.get_data_func("reed_distance", 2, 0.001), data_field_coord(3, 1)),
+
+            CentrePower(self.width, self.height),
+
+            Message(),
+        ]
+
+    def on_connect(self, client, userdata, flags, rc):
+        print('Connected with rc: {}'.format(rc))
+
+        for drawable in self.drawables:
+            drawable.draw_base(self.base_canvas)
+
+    def update_data_layer(self):
+        self.data_canvas.clear()
+
+        for drawable in self.drawables:
+            drawable.draw_data(self.data_canvas, self.data)
+
+    def get_data_func(self, data_key: str, decimals=0, scalar=1) -> Callable[[Data], str]:
```

Comment:
This is also useful for the base class too!

Reply:
I think it's only really going to be useful if we standardise the components and use them elsewhere. Otherwise, less complex overlays can just read `data["data_field_name"]` directly.

Reply:
Perhaps if we do end up using this everywhere then we should put this out!

Reply:
Since I've moved the components out, I may as well move this out too :)